### PR TITLE
Twins

### PR DESCRIPTION
**Pull request by sheharbano**
## Motivation

This PR adds the following Twins-related modifications:

- Increase round_initial_timeout for twins testing from default 1s to 2s
- Sort nodes in 'drop_config_test' and use await with commit_cb_receiver.next

This PR adds the following Twins-related functionality:

- Support dropping messages per round (DropConfigRound)
- Implement 'get_message_round' to extract round of a given message
- Support RoundProposers in 'start_num_nodes_with_twins'

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

- End-to-end twins tests:
'twins_proposer_test', 'twins_commit_test', 'twins_drop_config_round_test'

- Tests to check new DropConfigRound data structure:
'test_split_network_round', 'test_drop_message_for_round'

## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/libra/website, and link to your PR here.)

### CODE REVIEW

File: consensus/src/twins_test.rs [hunk 1]
```diff
@@ -353,18 +403,174 @@ fn twins_vote_dedup_test() {
...
+/// Let n0 (and implicitly twin0) be proposers
+///
+/// Test:
+///
+/// Extract enough votes so nodes in both partitions form commits.
+/// The commits should be on two different blocks
+///
+/// Run the test:
+/// cargo xtest -p consensus twins_proposer_test -- --nocapture
+#[cfg(test)]
```
Comment (zekun000) [inline consensus/src/twins_test.rs:438]: remove this line

File: consensus/src/network_tests.rs [hunk 2]
```diff
@@ -295,6 +304,23 @@ impl NetworkPlayground {
         msg_copies
     }
 
+    /// Return the round of a given message
+    fn get_message_round(&self, msg: ConsensusMsg) -> Option<u64> {
+        match msg {
+            ConsensusMsg::ProposalMsg(proposal_msg) => {
+                let unboxed = *proposal_msg;
```
Comment (zekun000) [inline consensus/src/network_tests.rs:311]: I don't think you need this, rust has auto deref so you can directly do proposal_msg.proposal().round()

File: consensus/src/network_tests.rs [hunk 3]
```diff
@@ -339,6 +365,48 @@ impl NetworkPlayground {
...
+
+    /// Creates the given per round network partitions
+    pub fn split_network_round(
+        &mut self,
+        round_partitions: &HashMap<u64, Vec<Vec<TwinId>>>,
+    ) -> bool {
+        let mut ret = true;
+
+        for (round, partitions) in round_partitions.iter() {
+            for (idx_curr_partition, curr_partition) in partitions.iter().enumerate() {
```
Comment (zekun000) [inline consensus/src/network_tests.rs:392]: this loop is quite intense, I think it just tries to achieve 
```
for i in 0..partitions.len() {
  for j in i+1..partitions.len() {
      self.drop_message_for_round(partitions[i], partitions[j], *round);
      ..
  }
}
```

File: consensus/src/network_tests.rs [hunk 4]
```diff
@@ -416,6 +484,44 @@ impl DropConfig {
...
 
+/// Table of per round message dropping rules
+struct DropConfigRound(HashMap<u64, DropConfig>);
+
+impl DropConfigRound {
+    /// Check if the message from 'src' to 'dst' should be dropped in the given round
+    pub fn is_message_dropped(&self, src: &TwinId, dst: &TwinId, round: u64) -> bool {
+        let mut result = false;
+
+        if self.0.contains_key(&round) {
```
Comment (zekun000) [inline consensus/src/network_tests.rs:495]: self.0.get(&round).and_then(|config| config.get(src).and_then(|set| set.contains(dst)).or_else(false)

File: consensus/src/twins_test.rs (hunk 1 above)
Comment (sheharbano) [reply consensus/src/twins_test.rs:438]: Done (assuming you meant line#438 '#[cfg(test)]')

File: consensus/src/network_tests.rs (hunk 2 above)
Comment (sheharbano) [reply consensus/src/network_tests.rs:311]: Done

File: consensus/src/network_tests.rs (hunk 3 above)
Comment (sheharbano) [reply consensus/src/network_tests.rs:392]: Not quite..

File: consensus/src/network_tests.rs (hunk 4 above)
Comment (sheharbano) [reply consensus/src/network_tests.rs:495]: That looks better + concise, indeed! Done (with slight modification):

        self.0
            .get(&round)
            .and_then(|config| config.0.get(src)
                .and_then(|set| Some(set.contains(dst))))
            .or_else(|| {None})
            .is_some()

File: consensus/src/network_tests.rs (hunk 3 above)
Comment (zekun000) [reply consensus/src/network_tests.rs:384]: sorry I misunderstood the loop, so here's what I got, the first two loops can also be replaced by map
```
for i in 0..partitions.len() {
  for j in i+1..partitions.len() {
    partitions[i].iter().flat_map(|src| partitions[j].iter().map(move |dst| (src, dst))).for_each(|(src, dst)| {
      self.drop_message_for_round(src, dst, *round);
      self.drop_message_for_round(dst, src, *round);
    })
  }
}
```

Comment (zekun000) [reply consensus/src/network_tests.rs:384]:
```
    partitions
        .iter()
        .enumerate()
        .flat_map(|(i, v1)| {
            partitions.iter().skip(i + 1).flat_map(move |v2| {
                v1.iter()
                    .flat_map(move |src| v2.iter().map(move |dst| (src, dst)))
            })
        })
        .for_each(|(src, dst)| println!("{} {}", src, dst));
```

Comment (sheharbano) [reply consensus/src/network_tests.rs]: Nice, thanks! Done.

**Review by zekun000 — APPROVED** [review]
/land

**Pull request closed without merging**